miniapps/elasticity/kernels/elasticity_kernels.hpp

```diff
@@ -0,0 +1,325 @@
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#ifndef MFEM_ELASTICITY_KERNELS_HPP
+#define MFEM_ELASTICITY_KERNELS_HPP
+
+#include "kernel_helpers.hpp"
+#include "linalg/vector.hpp"
+
+using mfem::internal::tensor;
+using mfem::internal::make_tensor;
+
+namespace mfem
+{
+
+namespace ElasticityKernels
+{
+
+template <int d1d, int q1d, typename material_type> static inline
+void Apply3D(const int ne,
+             const Array<double> &B_,
+             const Array<double> &G_,
+             const Array<double> &W_,
+             const Vector &Jacobian_,
+             const Vector &detJ_,
+             const Vector &X_, Vector &Y_,
+             const material_type &material)
+{
+   constexpr int dim = 3;
+   KernelHelpers::CheckMemoryRestriction(d1d, q1d);
+
+   // Basis functions evaluated at quadrature points.
+   const tensor<double, q1d, d1d> &B =
+   make_tensor<q1d, d1d>([&](int i, int j) { return B_[i + q1d*j]; });
+
+   // Gradients of basis functions evaluated at quadrature points.
+   const tensor<double, q1d, d1d> &G =
+   make_tensor<q1d, d1d>([&](int i, int j) { return G_[i + q1d*j]; });
+
+   const auto qweights = Reshape(W_.Read(), q1d, q1d, q1d);
+   // Jacobians of the element transformations at all quadrature points in
+   // column-major layout q1d x q1d x q1d x sdim x dim x ne
+   const auto J = Reshape(Jacobian_.Read(), q1d, q1d, q1d, dim, dim, ne);
+   const auto detJ = Reshape(detJ_.Read(), q1d, q1d, q1d, ne);
+   // Input vector
+   // d1d x d1d x d1d x vdim x ne
+   const auto U = Reshape(X_.Read(), d1d, d1d, d1d, dim, ne);
+   // Output vector
+   // d1d x d1d x d1d x vdim x ne
+   auto force = Reshape(Y_.ReadWrite(), d1d, d1d, d1d, dim, ne);
+
+   MFEM_FORALL_3D(e, ne, q1d, q1d, q1d,
+   {
+      // shared memory placeholders for temporary contraction results
+      MFEM_SHARED tensor<double, 2, 3, q1d, q1d, q1d> smem;
+      // cauchy stress
+      MFEM_SHARED_3D_BLOCK_TENSOR(invJ_sigma_detJw, double, q1d, q1d, q1d, dim, dim);
+      // du/dxi
+      MFEM_SHARED_3D_BLOCK_TENSOR(dudxi, double, q1d, q1d, q1d, dim, dim);
+
+      const auto U_el = Reshape(&U(0, 0, 0, 0, e), d1d, d1d, d1d, dim);
+      KernelHelpers::CalcGrad(B, G, smem, U_el, dudxi);
+
+      MFEM_FOREACH_THREAD(qx, x, q1d)
+      {
+         MFEM_FOREACH_THREAD(qy, y, q1d)
+         {
+            MFEM_FOREACH_THREAD(qz, z, q1d)
+            {
+               auto invJqp = inv(make_tensor<dim, dim>(
+               [&](int i, int j) { return J(qx, qy, qz, i, j, e); }));
+
+               auto dudx = dudxi(qz, qy, qx) * invJqp;
+
+               auto sigma = material.stress(dudx);
```

Comment:
I would put a big comment like `THE Q FUNCTION ACTION IS HERE`. This would help someone following this kernel recipe to understand what is going on.